File: RandytheStudent V2.0/RandyTheStudent/RandyTheStudent/InfoCharts.m
//
//  InfoCharts.m
//  RandyTheStudent
//
//  Created by Enrique Galicia on 04/03/14.
//  Copyright (c) 2014 Enrique Galicia. All rights reserved.
//

#import "InfoCharts.h"

@implementation InfoCharts

@synthesize Rectangulos;

- (id)initWithCoder:(NSCoder *)aDecoder
{
    if (self = [super initWithCoder:aDecoder])
    {
        [self setMultipleTouchEnabled:NO];
        [self setBackgroundColor:[UIColor clearColor]];

        
    }
    return self;
    
}
- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setMultipleTouchEnabled:NO];
        
    }
    return self;
}



- (void)drawRect:(CGRect)rect
{
    NSArray *Campos=[Rectangulos objectForKey:@"Campos"];
    NSArray *Acumulado=[Rectangulos objectForKey:@"Acumulado"];
    NSArray *Campo=[Rectangulos objectForKey:@"Campo"];
     NSArray *Valor=[Rectangulos objectForKey:@"Valor"];
    
    //NSLog(@"Puntos%@",[NSValue valueWithCGRect:CGRectMake(100, 0, rect.size.width-100,rect.size.height)]);
    
    
    NSArray *viewstoremove=[self subviews];
    for (UILabel *v in viewstoremove) {
        [v removeFromSuperview];
    }
    float altototal=rect.size.height;
    float divisiones=altototal/[Campos count];
    float maximo=0.0;
    if ([Campos count]>0){
        for (int a=0; a<=[Campos count]-1; a++) {
            UILabel *Nombre=[[UILabel alloc]init];
            Nombre.text=[Campos objectAtIndex:a];
            Nombre.frame=CGRectMake(0, 0, 100, divisiones-10);
            Nombre.center=CGPointMake(50, divisiones/2+5+(a*divisiones));
            Nombre.minimumScaleFactor=0.5;
            Nombre.adjustsFontSizeToFitWidth=YES;
            [self addSubview:Nombre];
        }
    }
    if ([Acumulado count]>0){
        for (int a=0; a<=[Acumulado count]-1; a++) {
            if (maximo<[[Acumulado objectAtIndex:a]floatValue]) {
                maximo=[[Acumulado objectAtIndex:a]floatValue];
            }
        }
    }
    float pixval=(rect.size.width-100)/maximo;
    
    
    //Create Background
    if ([Acumulado count]>0){
        for (int a=0; a<=[Acumulado count]-1; a++) {
            
            UIBezierPath *path = [UIBezierPath bezierPath];
            [path setLineWidth:1];
            [self drawrectangles:path posicion:CGRectMake(100, (0+a*divisiones), 100+[[Acumulado objectAtIndex:a] floatValue]*pixval, divisiones+(a*divisiones))];
            //NSLog(@" %iY%f Acum %f",a,(0+a*divisiones),[[Acumulado objectAtIndex:a] floatValue]*pixval);
            
            
            UIColor *fillColor = [UIColor lightGrayColor];
            [fillColor setFill];
            UIColor *strokeColor = [UIColor blackColor];
            [strokeColor setStroke];
            [path closePath];
            [path fill];
            [path stroke];
        }
    }
    if ([Campo count]>0){
        for (int a=0; a<=[Campo count]-1; a++) {
            NSArray *campo=[Campo objectAtIndex:a];
            NSArray *valor=[Valor objectAtIndex:a];
            float base=100;
            float azul=1;
            float rojo=1;
            float verde=1;
            if ([campo count]>0) {
                for (int b=0; b<=[campo count]-1; b++) {
                    
                    UIBezierPath *path;
                    path = [UIBezierPath bezierPath];
                    [path setLineWidth:1];
                    
                    [self drawrectangles:path posicion:CGRectMake(base, (0+a*divisiones),base+[[valor objectAtIndex:b] floatValue]*pixval, divisiones+(a*divisiones))];
                    base=base+[[valor objectAtIndex:b] floatValue]*pixval;
                    
                    azul=azul-0.05;
                    if (azul<0) {
                        azul=azul+1;
                    }
                    verde=verde-0.10;
                    if (verde<0) {
                        verde=verde+1;
                    }
                    rojo=rojo-0.15;
                    if (rojo<0) {
                        rojo=rojo+1;
                    }
                   

                    UIColor *fillColor = [UIColor colorWithRed:rojo green:verde blue:azul alpha:1];

                    [fillColor setFill];
                    UIColor *strokeColor = [UIColor blackColor];
                    [strokeColor setStroke];
                    [path closePath];
                    [path fill];
                    [path stroke];
                    
                   
                    
                }
            }
            
            
        }
    }
   
}


-(void)drawrectangles:(UIBezierPath*)pato posicion:(CGRect)posicion{
    //NSLog(@"Puntos%@",[NSValue valueWithCGRect:posicion]);
    [pato moveToPoint:CGPointMake(posicion.origin.x,posicion.origin.y)];
    [pato addLineToPoint:CGPointMake(posicion.size.width, posicion.origin.y)];
    [pato addLineToPoint:CGPointMake(posicion.size.width,posicion.size.height)];
    [pato addLineToPoint:CGPointMake(posicion.origin.x, posicion.size.height)];
    
    
    
}
-(void)updateview{
    [self setNeedsDisplay];
}








@end

File: RandytheStudent V2.0/RandyTheStudent/RandyTheStudent/InfoCharts.swift
//
//  InfoCharts.swift
//  RandyTheStudent
//
//  Swift port of the original InfoCharts.m (Objective-C) — a simple
//  hand-rolled horizontal stacked-bar chart view, driven by the
//  dictionary shape Subtitulados.subgraf(...) produces (keys "Campos",
//  "Acumulado", "Campo", "Valor").
//
//  `@objc(InfoCharts)` is required (not just a nicety) — both
//  MainViewController_iPhone.xib and MainViewController_iPad.xib
//  reference this class by the bare name "InfoCharts" in their
//  customClass attribute, and that string lookup only succeeds if a
//  class is registered under that exact Objective-C runtime name.
//

import UIKit

@objc(InfoCharts)
class InfoCharts: UIView {

    @objc var rectangulos: NSDictionary?

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        isMultipleTouchEnabled = false
        backgroundColor = .clear
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
        isMultipleTouchEnabled = false
    }

    override func draw(_ rect: CGRect) {
        let campos = (rectangulos?["Campos"] as? [Any]) ?? []
        let acumulado = (rectangulos?["Acumulado"] as? [Any]) ?? []
        let campo = (rectangulos?["Campo"] as? [Any]) ?? []
        let valor = (rectangulos?["Valor"] as? [Any]) ?? []

        for v in subviews {
            v.removeFromSuperview()
        }

        let altototal = Float(rect.size.height)
        let divisiones = campos.isEmpty ? 0 : altototal / Float(campos.count)

        if !campos.isEmpty {
            for a in 0..<campos.count {
                let nombre = UILabel()
                nombre.text = "\(campos[a])"
                nombre.frame = CGRect(x: 0, y: 0, width: 100, height: CGFloat(divisiones - 10))
                nombre.center = CGPoint(x: 50, y: CGFloat(divisiones / 2 + 5 + Float(a) * divisiones))
                nombre.minimumScaleFactor = 0.5
                nombre.adjustsFontSizeToFitWidth = true
                addSubview(nombre)
            }
        }

        var maximo: Float = 0.0
        for a in acumulado {
            let v = numeric(a)
            if maximo < v { maximo = v }
        }
        let pixval = maximo == 0 ? 0 : (Float(rect.size.width) - 100) / maximo

        // Create background
        for a in 0..<acumulado.count {
            let path = UIBezierPath()
            path.lineWidth = 1
            drawrectangles(path, posicion: CGRect(
                x: 100,
                y: CGFloat(Float(a) * divisiones),
                width: CGFloat(100 + numeric(acumulado[a]) * pixval),
                height: CGFloat(divisiones + Float(a) * divisiones)
            ))

            UIColor.lightGray.setFill()
            UIColor.black.setStroke()
            path.close()
            path.fill()
            path.stroke()
        }

        for a in 0..<campo.count {
            guard let campoRow = campo[a] as? [Any], let valorRow = valor[a] as? [Any] else { continue }
            var base: Float = 100
            var azul: Float = 1
            var rojo: Float = 1
            var verde: Float = 1

            for b in 0..<campoRow.count {
                let path = UIBezierPath()
                path.lineWidth = 1

                let ancho = numeric(valorRow[b]) * pixval
                drawrectangles(path, posicion: CGRect(
                    x: CGFloat(base),
                    y: CGFloat(Float(a) * divisiones),
                    width: CGFloat(base + ancho),
                    height: CGFloat(divisiones + Float(a) * divisiones)
                ))
                base += ancho

                azul -= 0.05
                if azul < 0 { azul += 1 }
                verde -= 0.10
                if verde < 0 { verde += 1 }
                rojo -= 0.15
                if rojo < 0 { rojo += 1 }

                UIColor(red: CGFloat(rojo), green: CGFloat(verde), blue: CGFloat(azul), alpha: 1).setFill()
                UIColor.black.setStroke()
                path.close()
                path.fill()
                path.stroke()
            }
        }
    }

    private func drawrectangles(_ pato: UIBezierPath, posicion: CGRect) {
        pato.move(to: CGPoint(x: posicion.origin.x, y: posicion.origin.y))
        pato.addLine(to: CGPoint(x: posicion.size.width, y: posicion.origin.y))
        pato.addLine(to: CGPoint(x: posicion.size.width, y: posicion.size.height))
        pato.addLine(to: CGPoint(x: posicion.origin.x, y: posicion.size.height))
    }

    @objc func updateview() {
        setNeedsDisplay()
    }

    /// Mirrors Objective-C's lenient `-[NSString floatValue]` (parses a
    /// leading numeric prefix, 0 if none) rather than Swift's strict
    /// `Float(String)` (nil unless the whole string parses).
    private func numeric(_ value: Any) -> Float {
        ("\(value)" as NSString).floatValue
    }
}
